### PR TITLE
QUIC server: Spawn for each stream

### LINKED ISSUE #26031: QUIC server introduces head-of-line blocking for a connection with multiple streams
#### Problem
The quic server supports multiple streams for a single connection but handles streams serially. This means that a transaction packet sent with stream 2 cannot be handled until the transaction packet sent with stream 1 has been fully received. This will decrease throughput and lessens the effectiveness of using multiple streams per connection.

We block to read from each stream here:
https://github.com/solana-labs/solana/blob/62ff54d04ede19cd6f8bcfe7c578a1297c41c0e3/streamer/src/nonblocking/quic.rs#L179

Here's a doc comment in quinn that advises against the approach we're using:
https://github.com/quinn-rs/quinn/blob/cd00119d254f9442618a7f1b8f748dcb9f309740/quinn/src/connection.rs#L567-L576

#### Proposed Solution
Don't use await for uni-stream read futures

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

awaiting streams in order can produce head of line blocking when streams come in out of order

#### Summary of Changes

spawn for each stream

Fixes #26031

### CODE REVIEW

**Comment by sakridge** [comment]
bench results:
this PR:
```
[2022-06-25T10:43:31.843037850Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 2 threads: kpps: 27.40 mbps: 5.48
     send_streams took 7.3s received_packets took 7.3s
[2022-06-25T10:43:36.183029597Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 4 threads: kpps: 46.08 mbps: 9.22
     send_streams took 4.3s received_packets took 4.3s
[2022-06-25T10:43:38.993553940Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 8 threads: kpps: 71.16 mbps: 14.23
     send_streams took 2.8s received_packets took 2.8s
[2022-06-25T10:43:41.272188161Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 16 threads: kpps: 87.77 mbps: 17.55
     send_streams took 2.2s received_packets took 2.3s
[2022-06-25T10:43:43.195212551Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 32 threads: kpps: 104.00 mbps: 20.80
     send_streams took 1.9s received_packets took 1.9s
```

master:
```
[2022-06-25T10:48:47.868729557Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 2 threads: kpps: 26.19 mbps: 5.24
     send_streams took 7.6s received_packets took 7.6s
[2022-06-25T10:48:52.237211401Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 4 threads: kpps: 45.78 mbps: 9.16
     send_streams took 4.3s received_packets took 4.4s
[2022-06-25T10:48:54.843824592Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 8 threads: kpps: 76.73 mbps: 15.35
     send_streams took 2.6s received_packets took 2.6s
[2022-06-25T10:48:56.935298626Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 16 threads: kpps: 95.63 mbps: 19.13
     send_streams took 2.1s received_packets took 2.1s
[2022-06-25T10:48:58.771416279Z INFO  solana_streamer::quic::test] sent 200000x 200b packets with 32 threads: kpps: 108.93 mbps: 21.79
     send_streams took 1.8s received_packets took 1.8s
[2022-06-25T10:48:59.260755423Z INFO  solana_streamer::quic::test] after store took 489ms
```

**Comment by sakridge** [comment]
This PR:
spawn-every-stream-bench-tps

Both compared, first master then this PR:
first-master-2nd-spawn-every-stream

**Comment by pgarg66** [comment]
Do you think usage of `tokio::time:timeout` solved this problem already? Specifically this code in `quic.rs`

```
                            if let Ok(chunk) = tokio::time::timeout(
                                Duration::from_millis(WAIT_FOR_STREAM_TIMEOUT_MS),
                                stream.read_chunk(PACKET_DATA_SIZE, false),
                            )
                            .await
```

**Comment by lijunwangs** [comment]
> This PR: spawn-every-stream-bench-tps
> 
> Master: bench-tps-master
> 
> Both compared, first master then this PR: first-master-2nd-spawn-every-stream

This seems to make the transaction throughput and confirmation time more evenly distributed over time compared with master?

File: streamer/src/nonblocking/quic.rs [hunk 1]
```diff
@@ -237,33 +237,39 @@ async fn handle_connection(
...
-                                    stats.clone(),
-                                    stake,
-                                ) {
-                                    last_update.store(timing::timestamp(), Ordering::Relaxed);
-                                    break;
+                        let stream_exit = stream_exit.clone();
+                        let stats = stats.clone();
+                        let packet_sender = packet_sender.clone();
+                        let last_update = last_update.clone();
+                        tokio::spawn(async move {
```
Comment (lijunwangs) [inline streamer/src/nonblocking/quic.rs:244]: The concern probably is we may spawn too many async tasks into the runtime in the server side compared with before. But I think that might be addressed by tweaking the maximum streams allowed for each connection if we find issues.

**Review by lijunwangs — APPROVED** [review]

**Review by pgarg66 — APPROVED** [review]

**Pull request merged**